### PR TITLE
[No QA] Use full Git history to populate StagingDeployCash

### PR DESCRIPTION
**Pull request by roryabraham**
### Details
The main motivation behind this pull request is to use the full Git / GitHub history to populate the StagingDeployCash, rather than relying on one-off changes made in actions. This makes the whole process more robust: any time the StagingDeployCash is updated, it should have all the correct pull requests and deploy blockers. That means if there is a GH outage, then when the service is restored the StagingDeployCash should self-resolve to its correct state.

Here's a more detailed list of the changes made in this PR (in broad strokes)

1. Changes `GitUtils.getPullRequestsMergedBetween` to be synchronous instead of async. Before it was unnecessarily async, and an arbitrary source of extra complexity.
1. Removes `createNewStagingDeployCash` and `updateStagingDeployCash` from the `GithubUtils` class. This is now handled entirely in the `createOrUpdateStagingDeploy` GitHub Action (located at `.github/actions/createOrUpdateStagingDeploy/createOrUpdateStagingDeploy.js`). Accordingly, I stripped some tests from `GitHubUtilsTest` and created a new automated test for the `createOrUpdateStagingDeploy` GH Action.
1. Removes the `NEW_PULL_REQUESTS` and `NEW_DEPLOY_BLOCKERS` inputs from the `createOrUpdateStagingDeploy` action. Instead, the action finds pull requests by looking at the tag on the previous StagingDeployCash, and parsing the git history between that tag and the latest tag to determine which PRs were merged since the last prod deploy. It finds deploy blockers by just querying the GH api for any open issues with the `DeployBlockerCash` label.
1. Updates the YML workflows to remove those unused inputs from usages of the `createOrUpdateStagingDeploy` action.

### Fixed Issues
Fixes https://github.com/Expensify/Expensify/issues/164517

### Tests
1. Merge this pull request.
1. Verify that a staging deploy happens and that the `StagingDeployCash` is updated with this pull request and the new `BUILD` version.
1. Check off this pull request on the `StagingDeployCash`
1. Lock the `StagingDeployCash`, then immediately after merge a pull request.
1. Verify that the `StagingDeployCash` is updated to include that pull request.
1. Merge another pull request.
1. Verify that the the pull request gets the "deploy deferred" comment.
1. Verify that the pull request _is not_ deployed to staging and _is not_ added to the `StagingDeployCash`
1. Create a new issue labelled `DeployBlockerCash`
1. Verify that the `StagingDeployCash` is updated to include the new deploy blocker
1. Label this pull request with `DeployBlockerCash`
1. Verify that the `StagingDeployCash` is updated to list this pull request as a deploy blocker.
1. Unlock and relock the `StagingDeployCash`.
1. Verify that the `StagingDeployCash` is update to include the deferred pull request.
1. Verify that the items on the `StagingDeployCash` that were checked off remain checked off.
1. Merge another pull request. Verify that it gets the "deploy deferred" comment and is not deployed to staging / added to the `StagingDeployCash`.
1. Close the `StagingDeployCash` to trigger a production deploy. Verify that the prod deploy begins.
1. Verify that a new `StagingDeployCash` is created, including the second deferred PR.
1. Once the prod deploy completes, verify that every PR on the previous `StagingDeployCash` gets the "deployed to prod" comment.

### CODE REVIEW

**Comment by roryabraham** [comment]
HOLD for live testing

**Comment by roryabraham** [comment]
@Dal-Papa All the `index.js` files here are compiled, so when you review this it's best just to hide them, as they'll be look funky and it will appear to be a lot of duplicate code

File: .github/actions/createOrUpdateStagingDeploy/createOrUpdateStagingDeploy.js [hunk 1]
```diff
@@ -4,66 +4,147 @@ const moment = require('moment');
...
-    .then(({data}) => console.log('Successfully created new StagingDeployCash! 🎉', data.html_url))
-    .catch((err) => {
-        console.error('An error occurred!', err);
-        core.setFailed(err);
-    });
+        })
+        .then(({data}) => {
+            console.log(
+                `Successfully ${shouldCreateNewStagingDeployCash ? 'created new' : 'updated'} StagingDeployCash! 🎉`,
+                data.html_url,
```
Comment (Dal-Papa) [inline .github/actions/createOrUpdateStagingDeploy/createOrUpdateStagingDeploy.js:136]: Why the mix between template strings and `console.log` arguments ?

**Review by Dal-Papa — COMMENTED** [review]
Asking this question now even though I'm not done with the review to avoid losing too much time.

**Comment by Dal-Papa** [comment]
> @Dal-Papa All the `index.js` files here are compiled, so when you review this it's best just to hide them, as they'll be look funky and it will appear to be a lot of duplicate code

Just to make sure, when you say compiled, that means that there are multiple source files that end up being the various `index.js` files right ? I see a lot of duplicate code but it's because of that, right ?

File: .github/actions/createOrUpdateStagingDeploy/createOrUpdateStagingDeploy.js (hunk 1 above)
Comment (roryabraham) [reply .github/actions/createOrUpdateStagingDeploy/createOrUpdateStagingDeploy.js:136]: Meh, no method to the madness here. This was prettymuch an arbitrary decision, but I can clean it up!

**Comment by roryabraham** [comment]
> Just to make sure, when you say compiled, that means that there are multiple source files that end up being the various index.js files right 

Yeah, that's correct. A more accurate word than "compiled" would be "bundled with dependencies". Basically any dependencies are bundled into a single Node.js script that has everything it needs to run. There's some more information [here](https://github.com/Expensify/Expensify.cash/blob/main/.github/workflows/README.md#why-do-actions-need-to-be-compiled)

File: .github/actions/createOrUpdateStagingDeploy/createOrUpdateStagingDeploy.js (hunk 1 above)
Comment (roryabraham) [reply .github/actions/createOrUpdateStagingDeploy/createOrUpdateStagingDeploy.js]: Updated 👍

**Review by Dal-Papa — DISMISSED** [review]
It looks good but there's a problem with that validate action ?

**Comment by roryabraham** [comment]
Ah yes, forgot to rebuild the GHA's 🤦

**Review by Dal-Papa — APPROVED** [review]
✅

**Comment by roryabraham** [comment]
Okay, making a note here that during our testing of CP's [over here](https://github.com/Expensify/Expensify.cash/pull/2978) we discovered another race condition:

1. The `StagingDeployCash` is locked, and starts making a new `PATCH` version./
2. Another PR is merged to `main`. Since the `StagingDeployCash` is locked, it does not create a `BUILD` version or start a staging deploy. It gets the "deploy deferred" comment.
3. The `PATCH` version finishes and is merged to `main`.
4. `staging` is updated from `main` 💥 – this includes the last-minute PR that was merged while the `StagingDeployCash` was locked.
5. So the last-minute PR is deployed, but isn't included in the `StagingDeployCash` for QA.

Fortunately, the output of this command:

```bash
git log --format="%s" 1.0.56-4...1.0.57-0
```

Includes that last-minute PR. So after the changes from this PR, it would be included in the `StagingDeployCash`, and not skipped for QA 👍

The last-minute PR should also get the "deployed to staging" comment, so that's good. The only thing that's sort of broken here is that this comment is a bit misleading:

image

But that's not a severe enough problem to worry about for such a rare edge-case.

**Comment by roryabraham** [comment]
Updated the testing steps of this PR to verify that it fixes this edge case 👍

**Review by AndrewGable — APPROVED** [review]

**Comment by AndrewGable** [comment]
Looks good, @roryabraham feel free to self merge when ready to test.

**Comment by roryabraham** [comment]
Thanks, I have to wait to test this until after the next production deploy.

**Comment by roryabraham** [comment]
Going to test this now

**Comment by roryabraham** [comment]
This did not work. First follow-up [here](https://github.com/Expensify/Expensify.cash/pull/3642) just adds better logs to help me diagnose [the failure](https://github.com/Expensify/Expensify.cash/runs/2851188746?check_suite_focus=true)

**Comment by roryabraham** [comment]
Think I found the problem and have a solution [here](https://github.com/Expensify/Expensify.cash/pull/3647)

**Pull request merged**